### ISSUE TITLE
Investigate whether the table can be left in a "good" state after uthash_fatal

### ISSUE THREAD

**Issue by Quuxplusone** [troydhanson/uthash#93]
Basically I'd like to either explicitly support (and unit-test) cases like the following, or explicitly document that these cases are not supported.  But if we do the latter, we'll be explicitly "unsupporting" some code that's already out there in the wild (see #92).

```
#include "uthash.h"
#undef uthash_fatal

#undef uthash_malloc
void *uthash_malloc(size_t sz) {
    static int count = 0;
    if (++count == FAIL_ON_THIS_ALLOCATION) return NULL;
    return malloc(sz);
}

struct Table {
    int hashkey;
    UT_hash_handle hh;
};

bool addItemOrFail(struct Table *t, int newValue) {
    struct Table *item = malloc(sizeof *item);
    item->hashkey = 42;
#define uthash_fatal(msg) return false
    HASH_ADD_INT(table, hashkey, item);
#undef uthash_fatal
    return true;
}

int main() {
    struct Table *t = NULL;
    addItemOrFail(t, 41);    // Suppose this fails.
    addItemOrFail(t, 42);    // Is this undefined behavior, or does it Just Work?
    addItemOrFail(t, 43);
    HASH_CLEAR(hh, t);    // Is this undefined behavior, or does it Just Work?
}
```

**Comment by Quuxplusone** [comment]
@tbeu: Bah, I wrote a big long response on https://trac.modelica.org/Modelica/ticket/2097 and then the spam filter rejected it. Long story short, watch *this* GitHub issue for updates on the definedness of your approach. I now think Modelica should simply `abort()` on malloc failure instead of trying to be clever with control flow.

Either way, I'd strongly prefer if you credited yourself in the Modelica changelog, instead of me, to avoid people thinking they should contact me about Modelica issues just because my name appears high up in the changelog. Thanks!

**Comment by tbeu** [comment]
Any news on this?

**Comment by Quuxplusone** [comment]
Sorry, no news is bad news in this case. I did maybe half of the investigation I was planning to do, and the results look very bad so far.

- The good news: `uthash_fatal` is used *only* to recover from `uthash_malloc` failure. Every call to `uthash_malloc` is paired with a conditional call to `uthash_fatal`.

- The bad news: One of those malloc calls is located [inside `HASH_EXPAND_BUCKETS`](https://github.com/troydhanson/uthash/blob/master/src/uthash.h#L799), which is [called from `HASH_ADD_TO_BKT`](https://github.com/troydhanson/uthash/blob/master/src/uthash.h#L745), which is [called from `HASH_ADD_KEYPTR_BYHASHVALUE_INORDER`](https://github.com/troydhanson/uthash/blob/master/src/uthash.h#L265). If this `uthash_fatal` returns, then the currently-being-added item will be in the hash table but not in the bloom filter, which is a bad state.

- More bad news: `HASH_CLEAR` [assumes](https://github.com/troydhanson/uthash/blob/master/src/uthash.h#L977) that `head != NULL` implies `(head)->hh.tbl != NULL`. This would be false if [the first allocation in `HASH_MAKE_TABLE`](https://github.com/troydhanson/uthash/blob/master/src/uthash.h#L176) had failed.

- We could make `HASH_MAKE_TABLE` recover from malloc failure by setting `(head)=NULL;`, but the `HASH_ADD` family [currently assumes](https://github.com/troydhanson/uthash/blob/master/src/uthash.h#L263) that falling out the bottom of `HASH_MAKE_TABLE` implies `head != NULL`.

**Comment by tbeu** [comment]
Thanks for your thorough investigations. Looks like we need to hope that malloc will never fail.

**Comment by Quuxplusone** [comment]
No new information, just summarizing:

- If `uthash_fatal` falls out the bottom, the behavior is undefined (in that your program will almost certainly crash).
- If `uthash_fatal` executes a`break` or `continue`, the behavior is undefined (in that your program will almost certainly crash).
- If `uthash_fatal` executes a `return`, control flow will be intact, but the hash table may be left in a state that cannot be cleaned up via any existing uthash macros.
- If `uthash_fatal` executes a `longjmp`, or `throw`, control flow will be intact, but the hash table may be left in a state that cannot be cleaned up via any existing uthash macros.
- If `uthash_fatal` terminates the program, everything's peachy.

One possible improvement to the status quo would be for us to add a macro `HASH_RECOVER_AFTER_FATAL(head, hh)` with the idea that you might call this macro from within your `longjmp` handler in order to get memory back in a consistent state. For example, it might do

    if (!(head) || !(head)->hh.tbl) {
        (head) = NULL;
    } else if (!(head)->hh.tbl->buckets) {
        uthash_free((head)->hh.tbl, sizeof(UT_hash_table));
        (head) = NULL;
    } else if (!(head)->hh.tbl->bloom_bv) {
        uthash_free((head)->hh.tbl->buckets, (head)->hh.tbl->num_buckets*sizeof(struct UT_hash_bucket));
        uthash_free((head)->hh.tbl, sizeof(UT_hash_table));
        (head) = NULL;
    } else if (...

But I haven't really thought that through 100%, especially for objects with non-trivial "destructors" (i.e. ones for which calling `HASH_CLEAR` on a non-empty table would leak resources).

**Comment by veselov** [comment]
This (graceful recovery from ENOMEM) would be quite an important feature. I'm working on a library that needs to return memory failure errors in case there wasn't any more left. Crashing (exiting) the app is not really an option (the app can do that if it wants to, but not the lib).

**Comment by veselov** [comment]
It seems there are a few things to do here.
1) enable `uthash_fatal` to be a block that is executed indicating an out-of-memory condition is encountered. May be even rename it it to `uthash_oom`.
2) Don't wrap macros that invoke `uthash_fatal` in `do{...}while(0)`, instead, place `do{}while(0)` at the end, to avoid creating a nested loop that the caller has to work around (if using `break`/`continue` in the oom handler). If variables are created in a define, enclose block into `{}` 
3) When an inner macro fails to allocate, have the macro stop, and mark the target memory as 0, or raise an allocation failure flag. The higher level macro may conditionally continue or clean up based on checking the value/flag.
4) Add additional reporting to the user. Actually, this can be an alternative to messing with providing non-destructive `uthash_fatal` implementation. To preserve existing behavior, check if the `uthash_fatal` is literal `CONT` (for example), and then as a failure path, set some flag in `hh` structure indicating that the operation has failed. The code that expects a potential failure will handle this just perfectly, without having to deal with potentially tricky control flow mechanism implemented in `uthash_fatal`.
5) HASH_SELECT will probably suffer, but IMO, it should keep going, raising "failed" flag in any selected items that failed to insert.

Would you be interested in a PR attempt?

**Comment by Quuxplusone** [comment]
Sorry, I think "2." is a non-starter. Every macro in uthash is ["hygienic"](https://en.wikipedia.org/wiki/Hygienic_macro), meaning that it behaves either like an expression or like a statement (via wrapping in `do {} while (0)`). Unwrapping any of these macros would cause client code such as

    if (cond)
        HASH_CLEAR(hh, myhash);
    else
        do_something_else();

to misbehave.

I think the goal of "3." is desirable, but the PR might be *very* messy, and the resulting code (with lots of duplicated error-checking conditions) might be less performant than the current approach, which I think might be a problem for some client code. I certainly won't stop you from trying.

"1." is certainly a good idea, modulo backward-compatibility concerns, but I believe it is completely orthogonal to the rest of your ideas (because it is an axiom that `uthash_fatal` ought to be defined by the client code as a hygienic macro, and any single expression is hygienic, and any arbitrary block of code can be converted into an expression by putting the block into a function `foo` and making the expression `foo()`).  My comments on "1." and "2." are all related to the general idea of "Code should be factorable and composable/decomposable without having to think too hard about global side effects."

Re "4.", I'm not sure I understand the details, but I have no specific comments on anything you said there.

Re "5.", my only comment is that I personally don't understand the use-case for `HASH_SELECT` as opposed to a `HASH_ITER`-containing-a-conditional-`HASH_ADD`. Is the *only* difference that `HASH_SELECT` doesn't repeat the debugging `HASH_FSCK` on every insert? If so, is that important enough for me to care about it? @troydhanson what do you think about `HASH_SELECT`?

**Comment by veselov** [comment]
"2." - conceded.

Regarding introducing more error checking. IMHO, good programming stack has no choice but to do that. A typical function that may return an error depends on the execution of some other functions and/or code that may fail, so it has to constantly check their result, and break out if they fail. The typical exceptions to that are functions that chain well even if invalid arguments are provided (e.g. you can call free(0) without checking for 0, and without adverse effects). Of course such practice will increase the size of the code, but IMO, the alternative of having code that either succeeds or terminates your program is worse. I can make that extra error checking to only kick in if the user defines a flag that enables so. After all, it only makes sense if the `uthash_fatal` does not terminate the execution thread, which is not a condition that would ever "work" before.

"4." only speaks to how to tell the caller that an operation fails. I don't want to rely on `uthash_fatal` function to have done it, because it may be cumbersome, i.e. having a flag in `hh` structure that says so is cleaner. I actually would like your feedback on it, because it will introduce such flag into `hh`.

**Comment by troydhanson** [comment]
@Quuxplusone, regarding `HASH_SELECT`, my preference would be to retire it altogether. The idea was to avoid recalculating the hash function if I recall correctly. It might have seemed useful to me at the time but in the years since I never found a need for it. I'd be fine with using ITER/ADD if I did.

Lots of good discussion here. @veselov I like the idea of 4. (And in general, I regret the fatal oom handling, sorry for the predicament!) With 4, existing code would work, but new code could define CONT as the fatal handler, and then the flag inside hh would indicate the error state (Right?) I have only given a quick perusal to this thread, and appreciate how much thought has gone into it. Off the top of my head it seems like the recover-after-fatal is still needed with 4, unless it did the malloc attempt before any manipulation of the hash table. I haven't looked at whether that's feasible.

**Comment by Quuxplusone** [comment]
Fixed by 01e5a878d7660e07177a7cb056c769297374d9e3!
